crypto/evp/keymgmt_lib.c

```diff
@@ -270,17 +277,33 @@ int evp_keymgmt_util_match(EVP_PKEY *pk1, EVP_PKEY *pk2, int selection)
          */
         if (keymgmt2 != NULL
             && keymgmt2->match != NULL) {
-            tmp_keydata = evp_keymgmt_util_export_to_provider(pk1, keymgmt2);
-            if (tmp_keydata != NULL) {
+            int ok = (pk1->keydata == NULL);
+
+            if (!ok) {
+                tmp_keydata =
+                    evp_keymgmt_util_export_to_provider(pk1, keymgmt2);
+                if (tmp_keydata != NULL)
+                    ok = 1;
+            }
+
+            if (ok) {
```

Comment:
Not sure why you moved this - it was clearer before.

Reply:
`evp_keymgmt_util_export_to_provider()` returns NULL on error...  it would be bad if we made it return 'cause `pk1->keydata` is NULL (which isn't an error), so we need to mitigate for that case.

Would this be better?

``` C
            tmp_keydata = NULL;
            if (pk1->keydata == NULL
                || (tmp_keydata = evp_keymgmt_util_export_to_provider(pk1, keymgmt2)) != NULL) {
                keymgmt1 = keymgmt2;
                keydata1 = tmp_keydata;
            }
```

Reply:
yep..

Reply:
Cool, that's an easy fix.  And thanks for pointing it out, that *does* look much cleaner indeed

Reply:
Actually, no...  conditions have changed, so this code becomes a bit more complex.  I've changed it a little bit, but most of all have added commentary to explain it.